model/roles.go

```diff
@@ -354,6 +360,102 @@ func (m *RoleManifest) SelectInstanceGroups(roleNames []string) (InstanceGroups,
 	return results, nil
 }
 
+// validateVariableDescriptions tests whether all variables have descriptions
+func validateVariableDescriptions(roleManifest *RoleManifest) validation.ErrorList {
+	allErrs := validation.ErrorList{}
+
+	for _, variable := range roleManifest.Variables {
+		if variable.CVOptions.Description == "" {
+			allErrs = append(allErrs, validation.Required(variable.Name,
+				"Description is required"))
+		}
+	}
+
+	return allErrs
+}
+
+// validateSortedTemplates tests that all templates are sorted in alphabetical order
+func validateSortedTemplates(roleManifest *RoleManifest) validation.ErrorList {
+	allErrs := validation.ErrorList{}
+
+	previousKey := ""
+
+	for _, templateDef := range roleManifest.Configuration.Templates {
+		key := templateDef.Key.(string)
```

Comment:
I just realized… we don't have anything to validate that `templateDef.Key` is castable to a string (and that `.Value` is also a string), or that the instance-group variants have string keys (and possibly non-string values, so no check there).

Reply:
does yaml allow non-string keys?
.Value is a string because we don't allow constants in the global section.
Not sure if I understand correctly.

Reply:
```diff
diff --git a/test-assets/role-manifests/app/tor-missing-description.yml b/test-assets/role-manifests/app/tor-missing-description.yml
index 0caa290..4d3d971 100644
--- a/test-assets/role-manifests/app/tor-missing-description.yml
+++ b/test-assets/role-manifests/app/tor-missing-description.yml
@@ -23,6 +23,7 @@ configuration:
     properties.tor.hashed_control_password: '((={{ }}=)){{PELERINUL}}'
     properties.tor.hostname: '((FOO))'
     properties.tor.private_key: '((#BAR))((HOME))((/BAR))'
+    3: 5
 variables:
 - name: BAR
   options:
```
```
==> Testing 
?       github.com/SUSE/fissile [no test files]
--- FAIL: TestMandatoryDescriptions (0.06s)
panic: interface conversion: interface {} is int, not string [recovered]
        panic: interface conversion: interface {} is int, not string

goroutine 69 [running]:
testing.tRunner.func1(0xc4201a60f0)
        /home/marky/opt/golang/current/src/testing/testing.go:711 +0x5d9
panic(0xaf3620, 0xc42037cbc0)
        /home/marky/opt/golang/current/src/runtime/panic.go:491 +0x2a2
github.com/SUSE/fissile/model.(*InstanceGroup).calculateRoleConfigurationTemplates(0xc4202682a0)
        /home/marky/src/work/src/github.com/SUSE/fissile/model/instance_groups.go:485 +0x903
```

And I'm saying we don't have any sanity checking that `.Value` _is_ a string; I can do `foo: [ a, b, c ]` and cause a panic.

Reply:
Yes, I agree.
This has been always the case afaict. 
https://www.pivotaltracker.com/story/show/160305258

Reply:
No, previously the struct was a `map[string]string`, and there was no casting (so we can't cast it wrong).